packages/node-core/src/indexer/dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {EventEmitter2} from '@nestjs/event-emitter';
+import {NETWORK_FAMILY} from '@subql/common';
+import {IBlock} from '@subql/types-core';
+import fetch from 'cross-fetch';
+import {NodeConfig} from '../../configure';
+import {getLogger} from '../../logger';
+import {BlockHeightMap} from '../../utils/blockHeightMap';
+import {DictionaryResponse, IDictionary, IDictionaryCtrl} from './types';
+
+const logger = getLogger('DictionaryService');
+export abstract class DictionaryService<DS, FB, D extends IDictionary<DS, FB>> implements IDictionaryCtrl<DS, FB> {
+  protected _dictionaries: D[] = [];
+
+  protected _currentDictionaryIndex: number | undefined;
+  constructor(
+    protected chainId: string,
+    protected readonly nodeConfig: NodeConfig,
+    protected readonly eventEmitter: EventEmitter2
+  ) {}
+
+  abstract initDictionaries(): Promise<void>;
+
+  init(dictionaries: D[]): void {
+    this._dictionaries = dictionaries;
+  }
+
+  private getDictionary(height: number): D | undefined {
+    if (this._dictionaries.length === 0) {
+      logger.debug(`No dictionaries available to use`);
+      return undefined;
+    }
+    // If current dictionary is valid, use current one instead of find a dictionary
+    if (
+      this._currentDictionaryIndex !== undefined &&
+      this._dictionaries[this._currentDictionaryIndex].heightValidation(height)
+    ) {
+      return this._dictionaries[this._currentDictionaryIndex];
+    } else {
+      this.findDictionary(height);
+      if (this._currentDictionaryIndex === undefined) {
+        logger.warn(`No supported dictionary found`);
+        return undefined;
+      } else {
+        logger.debug(`Updated : current dictionary Index is ${this._currentDictionaryIndex}`);
+        return this._dictionaries[this._currentDictionaryIndex];
+      }
+    }
+  }
+
+  // Find the next valid dictionary
+  private findDictionary(height: number, skipDictionaryIndex: number[] = []) {
+    // remove dictionary not valid
+    this._dictionaries = this._dictionaries.filter((d) => d.heightValidation(height));
+    this._currentDictionaryIndex = this._dictionaries.findIndex((d) => d.metadataValid);
+  }
+
+  useDictionary(height: number): boolean {
+    return !!this.getDictionary(height);
+  }
+
+  /**
+   *
+   * @param dataSources
+   */
+
+  buildDictionaryEntryMap(dataSources: BlockHeightMap<DS[]>): void {
+    for (const dict of this._dictionaries) {
+      dict.updateQueriesMap(dataSources);
+    }
+  }
+
+  async scopedDictionaryEntries(
```

Comment:
I think this needs some error handling. Looking at fetch service, if there is an error then it seems like it would keep trying the same query again.

Dictionary implementations should have some handling to decide if they are still valid and can be used.

This should also be the point where it could work its way through other dictionaries